node-red-project/flows.json

```diff
@@ -2399,6 +2542,89 @@
         "y": 580,
         "wires": []
     },
+    {
+        "id": "677a57bbf3619854",
+        "type": "http request",
+        "z": "9269520ddb532b00",
+        "g": "b334a2375eb8cb42",
+        "name": "Send event to TB server",
+        "method": "use",
+        "ret": "txt",
+        "paytoqs": "ignore",
+        "url": "",
+        "tls": "",
+        "persist": false,
+        "proxy": "",
+        "insecureHTTPParser": false,
+        "authType": "",
+        "senderr": false,
+        "headers": [],
+        "x": 850,
+        "y": 1340,
+        "wires": [
+            [
+                "df6ac2f42ac6d8c7"
+            ]
+        ]
+    },
+    {
+        "id": "df6ac2f42ac6d8c7",
+        "type": "debug",
+        "z": "9269520ddb532b00",
+        "g": "b334a2375eb8cb42",
+        "name": "send telemetry event response",
+        "active": true,
+        "tosidebar": true,
+        "console": false,
+        "tostatus": false,
+        "complete": "payload",
+        "targetType": "msg",
+        "statusVal": "",
+        "statusType": "auto",
+        "x": 1150,
+        "y": 1340,
+        "wires": []
+    },
+    {
+        "id": "4bec488857c0553f",
+        "type": "function",
+        "z": "9269520ddb532b00",
+        "g": "b334a2375eb8cb42",
+        "name": "setup telemetry payload & URL",
+        "func": "// format the payload with a timestamp\nlet values = { ...msg.payload }\nmsg.payload = { \n    ts: Date.now(),\n    values: values\n    }\n\n// The server URL, port, and device access token must be stored in the config context\n// before this will work\nlet serverUrl = global.get(\"serverUrl\", \"config\")\nlet serverPort = global.get(\"serverPort\", \"config\")\nlet accessToken = global.get(\"accessToken\", \"config\")\nmsg.url = `${serverUrl}:${serverPort}/api/v1/${accessToken}/telemetry`\nmsg.method = \"POST\";\nreturn msg;",
```

Comment:
Regardings `ts`: My preference would be this `events` payload gets assigned either the `frame_time` or `start_time` in order of preference. Right now it is effectively `end_time` plus some [milli]seconds.

Regarding `value`: The events values look like they are coming through as a dissociated list of key:value pairs. They do need to be associated with each other to have the proper meaning. Perhaps nest them under an `event` key similar to this:
[Example 4](https://thingsboard.io/docs/reference/http-api/#telemetry-upload-api:~:text=%3Atrue%7D%5D-,Example%204,-.%20Publish%20data%20as). Publish data as an object with timestamp (telemetry timestamp will be used) using data from **[telemetry-data-with-ts.json](https://thingsboard.io/docs/reference/resources/telemetry-data-with-ts.json)** file.

Reply:
I will change `ts` per your request.

I don't understand what you mean about `value`.  The events values are _already_ associated as you request.  Here's how they are packaged in node-red:
```payload: object
ts: 1724080954784
values: object
    id: "1724080938.431405-h7rmj1"
    camera: "picam_h264"
    label: "person"
    sub_label: null
    top_score: 0.828125
    frame_time: 1724080949.159755
    start_time: 1724080938.431405
    end_time: 1724080954.423641
    entered_zones: "["zone_capture","zone_near","zone_radar","zone_far"]"
    direction_calc: "outbound"
    speed_calc: undefined
    location: "tm-location-001"
_msgid: "0d6ad2c95c115b89"
url: "tb.untower.online:80/api/v1/cg2hhqqapp6s9h7i5syx/telemetry"
method: "POST"
```

The ThingsBoard device telemetry screen shows the values separately, but under-the-covers it must associate them with each other.  A few days ago, I made a ThingsBoard dashboard that uses a table widget to show multiple values in columns, with events in rows. That would not be possible if the values are not internally associated, probably by the `msgId` field.

(I would include a link to that dashboard here, but right now I am unable to connect to ThingsBoard.)

Reply:
What do you mean by "merciless changes"?

Reply:
> I will change `ts` per your request.
> 
> I don't understand what you mean about `value`. The events values are _already_ associated as you request. Here's how they are packaged in node-red:
> That would not be possible if the values are not internally associated, probably by the `msgId` field.


The `event` values need to be associated on the `event_id` (the `id` attribute in the `events` payload).  The TB system-generated `_msgid` is a good idea, but it's not stable if we, say, add additional fields that are unrelated to this event; e.g. if we add raw radar readings to this message, those are not necessarily part of this event.

We would at least need a nested structure to encapsulate the event (this assumes a single event per message):
```
ts: Date.now()
values: object
  event: object
    id: "1724080938.431405-h7rmj1"
    camera: "picam_h264"
    label: "person"
    sub_label: null
    top_score: 0.828125
    frame_time: 1724080949.159755
    start_time: 1724080938.431405
    end_time: 1724080954.423641
    entered_zones: "["zone_capture","zone_near","zone_radar","zone_far"]"
    direction_calc: "outbound"
    speed_calc: undefined
    location: "tm-location-001"
  radar: object
    ...
```
Then `ts` would truly be when the message was sent, not the event recorded.



> What do you mean by "merciless changes"?

Just a turn-of-phrase. I'm going to move the location details to the `tmdb` database as telemetry, so those will no longer need to be stored in context, but read from the db. This will support tracking deployments over time.

Reply:
I changed the ts in the telemetry message to be the frame_start, and added location, device name, latitude and longtitude to it. (I did this before seeing your msg that  you will put the location details in the tmdb database).

Reply:
> I changed the ts in the telemetry message to be the frame_start, and added location, device name, latitude and longtitude to it. (I did this before seeing your msg that you will put the location details in the tmdb database).

No problem. I'll take a look at it.

Just update the "setup telemetry payload & URL" node to use a nested `event`. This code should do it:
```
// format the payload with a timestamp
let values = { ...msg.payload }
msg.payload = {
    ts: Date.now(),
    values: {
        event: values
    }
}
```